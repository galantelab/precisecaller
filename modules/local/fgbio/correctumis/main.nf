process FGBIO_CORRECTUMIS {
    tag "$meta.id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/87/87626ef674e2f19366ae6214575a114fe80ce598e796894820550731706a84be/data' :
        'community.wave.seqera.io/library/fgbio:2.4.0--913bad9d47ff8ddc' }"

    input:
    tuple val(meta), path(bam)
    path umi_file
    val max_mismatches
    val min_distance

    output:
    tuple val(meta), path("${prefix}.bam") , emit: bam
    path "versions.yml"                    , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}_correct_umis"

    def mem_gb = 8
    if (!task.memory) {
        log.info '[fgbio CorrectUmis] Available memory not known - defaulting to 8GB. Specify process memory requirements to change this.'
    } else if (mem_gb > task.memory.giga) {
        if (task.memory.giga < 2) {
            mem_gb = 1
        } else {
            mem_gb = task.memory.giga - 1
        }
    }

    if ("${bam}" == "${prefix}.bam") {
        error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
    }

    """
    fgbio \\
        -Xmx${mem_gb}g \\
        --tmp-dir=. \\
        --async-io=true \\
        CorrectUmis \\
        ${args} \\
        --input ${bam} \\
        --output ${prefix}.bam \\
        --umi-files ${umi_file} \\
        --max-mismatches ${max_mismatches} \\
        --min-distance ${min_distance}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fgbio: \$( echo \$(fgbio --version 2>&1 | tr -d '[:cntrl:]' ) | sed -e 's/^.*Version: //;s/\\[.*\$//')
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}_correct_umis"

    if ("${bam}" == "${prefix}.bam") {
        error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
    }

    """
    touch ${prefix}.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fgbio: \$( echo \$(fgbio --version 2>&1 | tr -d '[:cntrl:]' ) | sed -e 's/^.*Version: //;s/\\[.*\$//')
    END_VERSIONS
    """
}
