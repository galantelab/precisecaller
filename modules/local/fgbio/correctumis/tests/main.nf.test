nextflow_process {

    name "Test Process FGBIO_CORRECTUMIS"
    script "../main.nf"
    process "FGBIO_CORRECTUMIS"

    tag "modules"
    tag "modules_"
    tag "fgbio"
    tag "fgbio/correctumis"

    test("Non fully degenerate dataset - bam") {

        when {
            config "./nextflow.config"

            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(params.local_testdata_base_path + 'data/Homo_sapiens/Simulations/bam/non_fully_degenerate_umis.bam', checkIfExists: true)
                ]
                input[1] = file(params.local_testdata_base_path + 'data/Homo_sapiens/Simulations/txt/expected_umis.txt', checkIfExists: true)
                input[2] = params.max_mismatches
                input[3] = params.min_distance
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.bam[0][1]).name,
                    process.out.cram,
                    process.out.versions
                ).match() }
            )
        }

    }

    test("Non fully degenerate dataset - stub") {

        options "-stub"

        when {
            config "./nextflow.config"

            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(params.local_testdata_base_path + 'data/Homo_sapiens/Simulations/bam/non_fully_degenerate_umis.bam', checkIfExists: true)
                ]
                input[1] = file(params.local_testdata_base_path + 'data/Homo_sapiens/Simulations/txt/expected_umis.txt', checkIfExists: true)
                input[2] = params.max_mismatches
                input[3] = params.min_distance
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    file(process.out.bam[0][1]).name,
                    process.out.cram,
                    process.out.versions
                ).match() }
            )
        }

    }

}
