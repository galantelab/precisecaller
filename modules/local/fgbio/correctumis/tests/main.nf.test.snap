{
    "Non fully degenerate dataset - stub": {
        "content": [
            "test_correct_umis.bam",
            null,
            [
                "versions.yml:md5,691dc7f665ce7035bf5c5e108ae96e22"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-09T02:45:10.836249659"
    },
    "Non fully degenerate dataset - bam": {
        "content": [
            "test_correct_umis.bam",
            null,
            [
                "versions.yml:md5,691dc7f665ce7035bf5c5e108ae96e22"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-09T02:44:55.120537791"
    }
}